### ISSUE TITLE
New Rule: Enum entries for enums matching a configuration should be in alphabetical order

### ISSUE THREAD

**Issue by drawers** [detekt/detekt#6115]
Searched in issues and could not find this suggested previously.

## Expected Behavior of the rule
This rule defines a configuration:

```yaml
EnumEntriesAlphabeticalOrder:
  active: false
  includeAnnotated: ["com.example.annotations.Alphabetical"]
  includeInterfaces: ["com.example.interfaces.Alphabetical"]
  include: ["com.example.MyEnum"]
```
This configuration allows a project to mark enums that they wish to keep in alphabetical order. They can do this in the most generic way by specifying their own annotation (e.g., a `@Alphabetical` annotation within their own project). Detekt will scan the project and look for enums that have that annotation explicitly applied or applied through implementing an interface decorated with the annotation. It will report when these enums have entries that are not in alphabetical order.

Further options for configuration are listing FQN of interfaces that mark enums to be scanned. I am not sure about using regex or wildcard lists to configure this rule

#### With annotation in configuration

```yaml
EnumEntriesAlphabeticalOrder:
  active: true
  includeAnnotated: ["com.example.annotations.Alphabetical"]
```

```kotlin
import com.example.annotations.Alphabetical

@Alphabetical
enum class AnalyticsVariable { // non-compliant (explicitly annotated with configured annotation)
   USER_ID,
   ITEMS_IN_CART,   
}
```

```kotlin
import com.example.annotations.Alphabetical

@Alphabetical
enum class AnalyticsVariable { // compliant 
   ITEMS_IN_CART,
   USER_ID, 
}
```

```kotlin
import com.example.annotations.Alphabetical

@Alphabetical
interface Analytics

// non-compliant (implements interface decorated with @Alphabetical and entries out of order)
enum class AnalyticsVariable : Analytics {
   USER_ID,
   ITEMS_IN_CART,
}

// compliant
enum class AnalyticsVariable: Analytics {
   ITEMS_IN_CART,
   USER_ID
}
```

#### With interfaces
```yaml
EntriesAlphabeticalOrder:
  active: true
  includeInterfaces: ["com.example.interfaces.Alphabetical"]
```

```kotlin
interface Alphabetical
```

```kotlin
// non-compliant (implements interface in config)
enum class AnalyticsVariable : Alphabetical {
   USER_ID,
   ITEMS_IN_CART,
}

// compliant 
enum class AnalyticsVariable: Alphabetical {
   ITEMS_IN_CART,
   USER_ID, 
}
```

## Context
In large Android codebases, the file containing an enum can become chaotic. It's not often not clear for devs where to add their new constant - should it be at the end of the file (to maintain chronological order) or somewhere else? It's not right to say every enum should be in alphabetical order because there are some where the ordinal matters (e.g., an enum where we are using the enum `ordinal` or when we are parsing an Int into an enum). Keeping in alphabetical order allows quick scanning of a file to understand contents but it generates a lot of friction in PRs and it also generates a noisy diff when someone decides to reorder a chaotic enum file. It can also generate potential bugs when people sort in alphabetical order without considering whether this is appropriate

I have worked on codebases where keeping enums entries in alphabetical order was required of devs and we solved it by writing our own Android Lint rule similar to how this proposed rule works. I've also vaguely remember hearing about it in projects at Uber (a difference in style project between the Uber Eats and the Uber Taxi projects?).

**Comment by cortinico** [comment]
This is a great rule. Additionally, I would suggest we should probably broaden it to include also lists or other occurrencies where you want to keep things sorted alphabetically.

As long as it concerns enums, IMHO the best approach is to offer only an annotation that would be used to enforce sorting. I would not introduce the `includeInterfaces` and `include` personally.

**Comment by 3flex** [comment]
I also wonder if we should have a detekt annotations artifact which includes a default annotation for this?

There's not much value in introducing it for a single annotation, but I keep looking at Error Prone's Immutable rule and annotation and think it would be a good one to introduce for detekt.

**Comment by drawers** [comment]
Thank you @cortinico , I agree that having the config refer to an annotation is the best solution for providing metadata like whether a particular set of class members should be put in alphabetical order. 

Based on your comment, I think the config should look like this now:

```yaml
EnumEntriesInAlphabeticalOrder:
  active: false
  includeAnnotated: ["com.example.annotations.Alphabetical"]
```yaml

> I would suggest we should probably broaden it to include also lists or other occurrencies where you want to keep things sorted alphabetically.

I am worried about creating a single large rule with lots of edge cases and tests that would need to be split up as a breaking change. In particular, I am thinking that extending this rule to sealed classes would be slightly more difficult because declarations inside a sealed class are more complex.

Would you consider a PR that just adds enum support? Or would you prefer a more general solution right from the start?

**Comment by drawers** [comment]
>I also wonder if we should have a detekt annotations artifact which includes a default annotation for this?
>There's not much value in introducing it for a single annotation,

@3flex this seems like a good idea and it will open up avenues for new rules where targeting specific members using YAML and regex is too clumsy. 

It sounds like introducing this rule requiring the user to define their own annotation would be okay for now and then when there are more rules that function in this way, the extra annotations artifact can be considered. What do you think?

**Comment by marschwar** [comment]
Just to clarify. This rule would not do anything unless there is an annotation configured, correct? So if I wanted (for whatever reason) to have all enums sorted alphabetically I would have to add the annotation to every enum?

**Comment by drawers** [comment]
> Just to clarify. This rule would not do anything unless there is an annotation configured, correct? So if I wanted (for whatever reason) to have all enums sorted alphabetically I would have to add the annotation to every enum?

@marschwar that is correct. You would have to add the annotation to every enum you wanted the entries to be kept in alphabetical order. Or add the annotation to a super interface and cause the enums you are interested in to derive from that interface. For example, in a previous code base I worked in there were feature flag enums declared across > 40 modules. These all derived from a common interface and it was desirable to keep them in alphabetical order. So we added an annotation to the interface and added an Android Lint rule similar to the Detekt rule I am proposing.

I have found that in large code bases you often end up with sets of enums that are better to order alphabetically and some that aren't (where chronological or some other order is more important). If there is a strong use case for keeping every enum in a codebase in alphabetical order, then we can revise the design here.

**Comment by BraisGabin** [comment]
I like the idea of this rule. And the idea that not all the enums should not follow this. But I'm not sure if it fits correctly on the detekt's main rule set exactly for that reason.

About having an annotation artifact: I'm not agains but I don't think that a rule like this should make us create all that complexity.

**Comment by cortinico** [comment]
I'm against having an artifact just for an annotation frankly. To me this rule could work in two modes:
- Catch all the occurrencies so you can keep all your enum/lists sorted
- Catch only the annotated ones, here user has to create an annotation and provide the FQN of the annotation

**Comment by drawers** [comment]
@BraisGabin how do you feel about @cortinico 's suggestion of two modes? 

I think that the first mode (catch all occurrences) with opt out via local suppression makes the proposed rule fit better with the operation of the other rules in the main rule set.

The second mode (via an annotation that the user has to declare themselves) is a little bit inconsistent with other rules but if we make the first mode the default it will be more transparent to consumers. Like @3flex stated earlier, annotations can open up the avenue for other rules that could be useful (like checking for immutability https://errorprone.info/api/latest/com/google/errorprone/annotations/Immutable.html)

I honestly don't mind making the rule in a separate repo if you still don't think it fits. What do you think?
